pkg/engine/mutation.go

```diff
@@ -73,24 +58,31 @@ func Mutate(policyContext PolicyContext) (resp response.EngineResponse) {
 			continue
 		}
 
+		// operate on the copy of the conditions, as we perform variable substitution
+		copyConditions := copyConditions(rule.Conditions)
 		// evaluate pre-conditions
-		if !variables.EvaluateConditions(ctx, rule.Conditions) {
+		// - handle variable subsitutions
+		if !variables.EvaluateConditions(ctx, copyConditions) {
 			glog.V(4).Infof("resource %s/%s does not satisfy the conditions for the rule ", resource.GetNamespace(), resource.GetName())
 			continue
 		}
 
+		mutation := rule.Mutation.DeepCopy()
 		// Process Overlay
-		if rule.Mutation.Overlay != nil {
-			var ruleResponse response.RuleResponse
-			ruleResponse, patchedResource = mutate.ProcessOverlay(ctx, rule, patchedResource)
-			if ruleResponse.Success {
-				// - variable substitution path is not present
-				if ruleResponse.PathNotPresent {
-					glog.V(4).Infof(ruleResponse.Message)
-					resp.PolicyResponse.Rules = append(resp.PolicyResponse.Rules, ruleResponse)
-					continue
-				}
+		if mutation.Overlay != nil {
+			overlay := mutation.Overlay
+			// subsiitue the variables
+			var err error
+			if overlay, err = variables.SubstituteVars(ctx, overlay); err != nil {
+				// variable subsitution failed
+				ruleResponse.Success = false
+				ruleResponse.Message = err.Error()
+				resp.PolicyResponse.Rules = append(resp.PolicyResponse.Rules, ruleResponse)
+				continue
```

Comment:
Is there a way to identify what type of failure the rue is? As in #607, we'd like to report policy error if the referenced path not exist.

Reply:
With this implementation No. As both scenarios are error. 
With the limitation of the library being used for evaluating JMESPATH, path not found and if the value stored is nil will result in same error. So we do not have a difference between the 2 scenarios.
For issue #607, we have to identify a way to differentiate between the 2 scenarios and then report errors of different types.

Reply:
It would be good to have a way to differentiate between `policy application failure` and `variable substitution failure - both path not found and if the value stored is nil`.

Reply:
True, but this is a limitation of the library being used by us. 
If we want to support the above, we'll have to fork the library and make the corresponding changes in the JMESPATH.

Reply:
Regarding separating the error types, yes they can be classified into groups. Right now the error type doesn't matter, as they both are reported as errors.

Reply:
Would it be helpful to set it here? Later it may be confusing to identify the error types as it is referred by all 3 types of policies.

Reply:
It would be helpful in the future. We can work on a framework and add define how to differentiate different errors and also their behavior.